packages/@lwc/engine/src/framework/vm.ts

```diff
@@ -142,55 +136,48 @@ export function rerenderVM(vm: VM) {
     rehydrate(vm);
 }
 
-export function appendRootVM(vm: VM) {
-    runConnectedCallback(vm);
-    rehydrate(vm);
-}
-
 export function appendVM(vm: VM) {
+    runConnectedCallback(vm);
     rehydrate(vm);
 }
 
 // just in case the component comes back, with this we guarantee re-rendering it
 // while preventing any attempt to rehydration until after reinsertion.
-function resetComponentStateWhenRemoved(vm: VM) {
+function reset(vm: VM) {
     const { state } = vm;
     if (state !== VMState.disconnected) {
         const { oar, tro } = vm;
-        // Making sure that any observing record will not trigger the rehydrated on this vm
+        // Making sure that any observing record will not trigger the rehydration on this vm
         tro.reset();
         // Making sure that any observing accessor record will not trigger the setter to be reinvoked
         for (const key in oar) {
             oar[key].reset();
         }
-        runDisconnectedCallback(vm);
-        // Spec: https://dom.spec.whatwg.org/#concept-node-remove (step 14-15)
-        runShadowChildNodesDisconnectedCallback(vm);
-        runLightChildNodesDisconnectedCallback(vm);
+    }
+    if (isFalse(vm.isDirty)) {
```

Comment:
I think this condition is unnecessary, setting it to true is the same thing... without having to ask for the previous value.